src/rust/engine/process_execution/src/remote/streaming.rs

```diff
@@ -0,0 +1,590 @@
+use std::collections::BTreeMap;
+use std::sync::Arc;
+use std::time::Duration;
+
+use async_trait::async_trait;
+use bazel_protos::call_option;
+use bazel_protos::error_details::PreconditionFailure;
+use bazel_protos::remote_execution::{
+  Action, Command, ExecuteRequest, ExecuteResponse, ExecutedActionMetadata, WaitExecutionRequest,
+};
+use bazel_protos::status::Status;
+use bytes::Bytes;
+use concrete_time::TimeSpan;
+use futures::compat::{Future01CompatExt, Stream01CompatExt};
+use futures::future::{self as future03};
+use futures::{Stream, StreamExt};
+use hashing::{Digest, Fingerprint};
+use log::{debug, trace, warn};
+use protobuf::Message;
+use store::Store;
+
+use super::{
+  format_error, maybe_add_workunit, populate_fallible_execution_result, ExecutionError,
+  OperationOrStatus,
+};
+use crate::{
+  Context, FallibleProcessResultWithPlatform, MultiPlatformProcess, Platform, PlatformConstraint,
+  Process, ProcessMetadata,
+};
+
+/// Implementation of CommandRunner that runs a command via the Bazel Remote Execution API
+/// (https://docs.google.com/document/d/1AaGk7fOPByEvpAbqeXIyE8HX_A3_axxNnvroblTZ_6s/edit).
+///
+/// Results are streamed from the output stream of the Execute function (and possibly the
+/// WaitExecution function if `StreamingCommandRunner` needs to reconnect).
+///
+/// If the CommandRunner has a Store, files will be uploaded to the remote CAS as needed.
+/// Note that it does not proactively upload files to a remote CAS. This is because if we will
+/// get a cache hit, uploading the files was wasted time and bandwidth, and if the remote CAS
+/// already has some files, uploading them all is a waste. Instead, we look at the responses we
+/// get back from the server, and upload the files it says it's missing.
+///
+/// In the future, we may want to do some clever things like proactively upload files which the
+/// user has changed, or files which aren't known to the local git repository, but these are
+/// optimizations to shave off a round-trip in the future.
+#[derive(Clone)]
+pub struct StreamingCommandRunner {
+  metadata: ProcessMetadata,
+  platform: Platform,
+  store: Store,
+  headers: BTreeMap<String, String>,
+  channel: grpcio::Channel,
+  env: Arc<grpcio::Environment>,
+  execution_client: Arc<bazel_protos::remote_execution_grpc::ExecutionClient>,
+}
```

Comment:
Yeah, I will be doing the timeout work in the follow-on PR. The current form of the PR only has a deadline timeout.

Reply:
Also backoff is less of a useful concept in streaming mode because, so long as the connection remains open (subject to keep-alive considerations), we won't resend the request. If we get disconnected, we just reconnect using WaitExecution and wait again with an open connection.